redskyctl/internal/commands/generate/rbac.go

```diff
@@ -90,88 +96,174 @@ func (o *RBACOptions) Complete() {
+	// Add up all the objects and print them out
+	rbac := buildRBAC(roleRef, subject, rules, namespaces)
+	return o.Printer.PrintObj(rbac, o.Out)
+}
+
+func (o *RBACOptions) bindingTargets(experimentList *redskyv1alpha1.ExperimentList) (*rbacv1.RoleRef, *rbacv1.Subject, []string, error) {
+	// Create the role reference
+	roleRef := &rbacv1.RoleRef{APIGroup: "rbac.authorization.k8s.io", Kind: "Role", Name: o.Name}
+	if o.ClusterRole {
+		roleRef.Kind = "ClusterRole"
+	}
+	if roleRef.Name == "" {
+		roleRef.Name = "redsky-patching-"
+		if len(experimentList.Items) == 1 && experimentList.Items[0].Name != "" {
+			roleRef.Name += experimentList.Items[0].Name
 		} else if o.Filename == "-" {
-			// TODO This needs some type of uniqueness
-			clusterRole.Name = "redsky-patching-stdin"
+			roleRef.Name += fmt.Sprintf("stdin-%s", rand.String(5))
 		} else if o.Filename != "" {
-			// TODO This needs more clean up
-			clusterRole.Name = "redsky-patching-" + strings.ToLower(filepath.Base(o.Filename))
+			dir, suffix := filepath.Split(o.Filename)
+			suffix = strings.TrimSuffix(suffix, filepath.Ext(suffix))
+			if suffix == "experiment" {
+				suffix = filepath.Base(dir)
+			}
+			re := regexp.MustCompile("[^a-z0-9]+")
+			suffix = re.ReplaceAllString(strings.ToLower(suffix), "-")
+			roleRef.Name += suffix
 		} else {
-			// TODO This probably doesn't work with cluster roles
-			clusterRole.GenerateName = "redsky-patching-"
+			roleRef.Name += rand.String(5)
 		}
 	}
 
-	// Update the cluster role binding based on the cluster role name
-	clusterRoleBinding.Name = clusterRole.Name + "binding"
-	clusterRoleBinding.RoleRef.Name = clusterRole.Name
-
-	// Add rules from the experiment
-	rules, err := o.findRules(experiment)
+	// Create the subject using the namespace of the controller from the configuration
+	ctrl, err := config.CurrentController(o.Config.Reader())
 	if err != nil {
-		return err
+		return nil, nil, nil, err
 	}
-	for _, r := range rules {
-		clusterRole.Rules = mergeRule(clusterRole.Rules, r)
+	subject := &rbacv1.Subject{
+		Kind:      "ServiceAccount",
+		Name:      "default",
+		Namespace: ctrl.Namespace,
 	}
 
-	// Do not generate an empty cluster role
-	if len(clusterRole.Rules) == 0 {
-		return nil
-	}
+	// Namespaces
+	var namespaces []string
+	if !o.ClusterRole || !o.ClusterRoleBinding {
+		// Get the distinct list of namespaces from the experiments
+		ns := make(map[string]bool, len(experimentList.Items))
+		for i := range experimentList.Items {
+			ns[experimentList.Items[i].Namespace] = true
+		}
 
-	l := &corev1.List{
-		Items: []runtime.RawExtension{
-			{Object: clusterRole},
-			{Object: clusterRoleBinding},
-		},
+		// Try to provide an explicit default value (cstr.Namespace can still be "")
+		if ns[""] {
+			if cstr, err := config.CurrentCluster(o.Config.Reader()); err == nil {
+				delete(ns, "")
+				ns[cstr.Namespace] = true
+			}
```

Comment:
Ultimately a nit, but could we do something like 

```
	var defaultNamespace string
	if cstr, err := config.CurrentCluster(o.Config.Reader()); err == nil {
		defaultNamespace = cstr.Namespace
	}

	ns := make(map[string]struct{}, len(experimentList.Items))
	for _, experiment := range experimentList.Items {
		if experiment.Namespace == "" {
			experiment.Namespace = defaultNamespace
		}
		ns[experiment.Namespace] = struct{}{}
	}
```
Since we don't actually need the value part of the map, I think a `struct{}` would be more appropriate to indicate this is a just a placeholder value with no actual purpose.